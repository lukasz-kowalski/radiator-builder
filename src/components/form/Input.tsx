import { InputHTMLAttributes } from "react";

interface Props extends InputHTMLAttributes<HTMLInputElement> {
  label: string;
  error: string | undefined;
}

export default function Input(props: Props): React.JSX.Element {
  const { label, error, ...inputProps } = props;

  return (
    <div className="flex flex-col w-full">
      <label className="text-sm" htmlFor={inputProps.id}>
        {label}
      </label>
      <input
        {...inputProps}
        className="p-1 border-2 h-10"
        aria-invalid={!!error}
        aria-errormessage={`error-${inputProps.id}`}
      />
      <div className="h-[1rem]">
        <p id={`error-${inputProps.id}`} className="text-red-600 text-sm">
          {error}
        </p>
      </div>
    </div>
  );
}
