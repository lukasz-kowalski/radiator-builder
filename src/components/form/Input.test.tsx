import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Input from "@/components/form/Input";

describe("Input component", () => {
  it("should render the label correctly", () => {
    render(
      <Input label="Radiator length" id="radiatorLength" error={undefined} />
    );

    expect(screen.getByLabelText("Radiator length")).toBeInTheDocument();
  });

  it("should pass input props correctly", () => {
    render(
      <Input
        label="Radiator length"
        id="radiatorLength"
        type="number"
        value="123"
        onChange={() => {}}
        error={undefined}
      />
    );

    const input = screen.getByLabelText("Radiator length") as HTMLInputElement;
    expect(input.type).toBe("number");
    expect(input.value).toBe("123");
  });

  it("should display error message and sets proper aria attributes", () => {
    render(
      <Input
        label="Radiator length"
        id="radiatorLength"
        error="Field is required"
      />
    );

    const input = screen.getByLabelText("Radiator length");
    const errorMessage = screen.getByText("Field is required");

    expect(input).toHaveAttribute("aria-invalid", "true");
    expect(input).toHaveAttribute("aria-errormessage", "error-radiatorLength");
    expect(errorMessage).toBeInTheDocument();
  });

  it("should not display error if error is undefined", () => {
    const { container } = render(
      <Input label="Radiator length" id="radiatorLength" error={undefined} />
    );

    const errorMessage = container.querySelector("#error-radiatorLength");

    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toHaveAttribute("id", "error-radiatorLength");
  });

  it("should call onChange handler when typing", async () => {
    const handleChange = jest.fn();

    render(
      <Input
        label="Radiator length"
        id="radiatorLength"
        value=""
        onChange={handleChange}
        error={undefined}
      />
    );

    const input = screen.getByLabelText("Radiator length");
    await userEvent.type(input, "500");

    expect(handleChange).toHaveBeenCalledTimes(3);
  });

  it("should match snapshot", () => {
    const { container } = render(
      <Input
        label="Radiator length"
        id="radiatorLength"
        type="number"
        value="300"
        error="Required field"
        onChange={() => {}}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
