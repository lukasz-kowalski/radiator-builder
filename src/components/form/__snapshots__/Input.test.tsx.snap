// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col w-full"
  >
    <label
      class="text-sm"
      for="radiatorLength"
    >
      Radiator length
    </label>
    <input
      aria-errormessage="error-radiatorLength"
      aria-invalid="true"
      class="p-2 border-2 h-10 bg-white"
      id="radiatorLength"
      type="number"
      value="300"
    />
    <div
      class="h-[3rem]"
    >
      <p
        class="text-red-600 text-sm"
        id="error-radiatorLength"
      >
        Required field
      </p>
    </div>
  </div>
</div>
`;
